cmd/snap/main.go

```diff
@@ -34,9 +34,33 @@ type options struct {
 
 var optionsData options
 
-var parser = flags.NewParser(&optionsData, flags.HelpFlag|flags.PassDoubleDash)
+var parser *flags.Parser
+
+// cmdInfo holds information needed to call parser.AddCommand(...).
+type cmdInfo struct {
+	name, shortHelp, longHelp string
+	builder                   func() interface{}
+}
+
+// commands holds information about all non-experimental commands.
+var commands []cmdInfo
+
+// Parser creates and populates a fresh parser.
+// Since commands have local state a fresh parser is required to isolate tests
+// from each other.
+func Parser() *flags.Parser {
+	parser := flags.NewParser(&optionsData, flags.HelpFlag|flags.PassDoubleDash)
+	// Add all regular commands
+	for _, c := range commands {
+		if _, err := parser.AddCommand(c.name, c.shortHelp, c.longHelp, c.builder()); err != nil {
+			logger.Panicf("unable to add command %q: %v", c.name, err)
+		}
+	}
+	return parser
+}
 
 func init() {
+	parser = Parser()
```

Comment:
I "kept" it here to maintain similarity to what was there before (parser was initialized at a global scope). I can move it to main if you think that's better.